modules/templates/htmlrenderer.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package templates
+
+import (
+	"context"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/watcher"
+
+	"github.com/unrolled/render"
+)
+
+var rendererKey interface{} = "templatesHtmlRendereer"
+
+// HTMLRenderer returns the current html renderer for the context or creates and stores one within the context for future use
+func HTMLRenderer(ctx context.Context) (context.Context, *render.Render) {
+	rendererInterface := ctx.Value(rendererKey)
+	if rendererInterface != nil {
+		renderer, ok := rendererInterface.(*render.Render)
+		if ok && renderer != nil {
+			return ctx, renderer
+		}
+	}
+
+	if setting.IsProd {
+		log.Log(1, log.DEBUG, "Creating static HTML Renderer")
+	} else {
+		log.Log(1, log.DEBUG, "Creating auto-reloading HTML Renderer")
+	}
+
+	renderer := render.New(render.Options{
+		Extensions:                []string{".tmpl"},
+		Directory:                 "templates",
+		Funcs:                     NewFuncMap(),
+		Asset:                     GetAsset,
+		AssetNames:                GetTemplateAssetNames,
+		UseMutexLock:              !setting.IsProd,
+		IsDevelopment:             false,
```

Comment:
Why not use render's internal auto reload?

Reply:
As I say above it locks all the templates and recompiles every time a template is requested.

Take a look at `r.HTML()`:

https://github.com/unrolled/render/blob/ae0c1ea59847313997034918c584be92fdd593d8/render.go#L435-L444

```go
func (r *Render) HTML(w io.Writer, status int, name string, binding interface{}, htmlOpt ...HTMLOptions) error {
	// If we are in development mode, recompile the templates on every HTML request.
	r.lock.RLock() // rlock here because we're reading the hasWatcher
	if r.opt.IsDevelopment && !r.hasWatcher {
		r.lock.RUnlock() // runlock here because CompileTemplates will lock
		r.CompileTemplates()
		r.lock.RLock()
	}
	templates := r.templates
	r.lock.RUnlock()
```

Here, `r.hasWatcher == false` always because we have set `Asset` and  `AssetNames`,  thus `r.CompileTemplates()` is called.

We can see that there is no fsnotify.Watcher goroutine listed on the process stack, and we can prove that the `r.hasWatcher == false` by looking at `r.CompileTemplates()`:

https://github.com/unrolled/render/blob/ae0c1ea59847313997034918c584be92fdd593d8/render.go#L196

```go
func (r *Render) CompileTemplates() {
	if r.opt.Asset == nil || r.opt.AssetNames == nil {
		r.compileTemplatesFromDir()
		return
	}

	r.compileTemplatesFromAsset()
}
```

and noting that because `Asset` and `AssetNames` are not `nil` we use `r.compileTemplatesFromAsset()`

<details><summary>Then we can look at `r.compileTemplatesFromAsset()` and see it does not set the watcher.</summary>

https://github.com/unrolled/render/blob/ae0c1ea59847313997034918c584be92fdd593d8/render.go#L286

```go
func (r *Render) compileTemplatesFromAsset() {
	dir := r.opt.Directory
	tmpTemplates := template.New(dir)
	tmpTemplates.Delims(r.opt.Delims.Left, r.opt.Delims.Right)

	for _, path := range r.opt.AssetNames() {
		if !strings.HasPrefix(path, dir) {
			continue
		}

		rel, err := filepath.Rel(dir, path)
		if err != nil {
			panic(err)
		}

		ext := ""
		if strings.Contains(rel, ".") {
			ext = "." + strings.Join(strings.Split(rel, ".")[1:], ".")
		}

		for _, extension := range r.opt.Extensions {
			if ext == extension {
				buf, err := r.opt.Asset(path)
				if err != nil {
					panic(err)
				}

				name := (rel[0 : len(rel)-len(ext)])
				tmpl := tmpTemplates.New(filepath.ToSlash(name))

				// Add our funcmaps.
				for _, funcs := range r.opt.Funcs {
					tmpl.Funcs(funcs)
				}

				// Break out if this parsing fails. We don't want any silent server starts.
				template.Must(tmpl.Funcs(helperFuncs).Parse(string(buf)))
				break
			}
		}
	}
	r.lock.Lock()
	defer r.lock.Unlock()
	r.templates = tmpTemplates
}
```

</details>